modules/core/src/main/java/org/apache/ignite/internal/managers/communication/transmit/channel/TransmitAbstractChannel.java

```diff
@@ -0,0 +1,180 @@
+ *
+ *     <ul>
+ *         <li>For read operations over the {@link InputStream} or write operation through the {@link OutputStream}
+ *         the {@link Socket#setSoTimeout(int)} will be used and an {@link SocketTimeoutException} will be
+ *         thrown when the timeout occured.</li>
+ *         <li>To achive the file zero-copy {@link FileChannel#transferTo(long, long, java.nio.channels.WritableByteChannel)}
+ *         the {@link SocketChannel} must be used directly in the blocking mode. For reading or writing over
+ *         the SocketChannels, using the <tt>Socket#setSoTimeout(int)</tt> is not possible, because it isn't
+ *         supported for sockets originating as channels. In this case, the decicated wather thread must be
+ *         used which will close conneciton on timeout occured.</li>
+ *     </ul>
+ * </p>
+ */
+public abstract class TransmitAbstractChannel implements Closeable {
+    /** */
+    private static final int DFLT_IO_TIMEOUT_MILLIS = 5_000;
+
+    /** */
+    private static final String RESET_BY_PEER_MSG = "Connection reset by peer";
+
+    /** */
+    private static final String CLOSED_BY_REMOTE_MSG = "An existing connection was forcibly closed by the remote host";
+
+    /** */
+    private final SocketChannel channel;
+
+    /** */
+    protected final IgniteLogger log;
+
+    /** */
+    private final int timeoutMillis;
+
+    /**
+     * @param ktx Kernal context.
+     * @param channel Socket channel to upload files to.
+     */
+    protected TransmitAbstractChannel(
+        GridKernalContext ktx,
+        SocketChannel channel
+    ) throws IOException {
+        this(ktx, channel, DFLT_IO_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * @param ktx Kernal context.
+     * @param channel Socket channel to upload files to.
+     * @param timeout Read\write timeout.
+     * @param unit The {@link TimeUnit} of given <tt>timeout</tt>.
+     * @throws IOException If channel configuration fails.
+     */
+    protected TransmitAbstractChannel(
+        GridKernalContext ktx,
+        SocketChannel channel,
+        int timeout,
+        TimeUnit unit
+    ) throws IOException {
+        assert ktx != null;
+        assert channel != null;
+        assert unit != null;
+
+        this.channel = channel;
+        log = ktx.log(getClass());
+        timeoutMillis = timeout <= 0 ? 0 : Math.max((int)unit.toMillis(timeout), DFLT_IO_TIMEOUT_MILLIS);
+
+        // Timeout must be enabled prior to entering the blocking mode to have effect.
+        channel.socket().setSoTimeout(timeoutMillis);
+        channel.configureBlocking(true);
+    }
+
+    /**
+     * @return The timeout of read\write operations in milliseconds.
+     */
+    public int timeout() {
+        return timeoutMillis;
+    }
+
+    /**
+     * @param cause The original cause to throw.
+     * @return The new cause or the old one.
+     */
+    public IOException transformExceptionIfNeed(IOException cause) {
```

Comment:
This can be protected (and static?)

Reply:
I've marked this method protected, but not `static`, since it internally use configured socket instance to produce a detailed error message. Is it ok?